src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocLeadingAsteriskAlignCheck.java

```diff
@@ -0,0 +1,241 @@
+ * </li>
+ * </ul>
+ *
+ * @since 10.17.0
+ */
+@GlobalStatefulCheck
+public class JavadocLeadingAsteriskAlignCheck extends AbstractJavadocCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_KEY = "javadoc.asterisk.indentation";
+
+    /** Specifies the column number of starting block of the javadoc comment with tabs expanded. */
+    private int commentBeginColNumber;
+
+    /** Specifies the line number of starting block of the javadoc comment. */
+    private int commentBeginLineNumber;
+
+    /**
+     * Specifies the column number of the leading asterisk
+     * where tabs are converted to a single whitespace.
+     */
+    private int singleSpaceAsteriskColNumber;
+
+    @Override
+    public int[] getDefaultJavadocTokens() {
+        return new int[] {
+            JavadocTokenTypes.LEADING_ASTERISK,
+        };
+    }
+
+    @Override
+    public int[] getRequiredJavadocTokens() {
+        return getAcceptableJavadocTokens();
+    }
+
+    @Override
+    public void beginJavadocTree(DetailNode rootAst) {
+        // this method processes and sets information of starting javadoc tag.
+        final String startLine = getLines()[rootAst.getLineNumber() - 1];
+        commentBeginLineNumber = rootAst.getLineNumber();
+        commentBeginColNumber = CommonUtil.lengthExpandedTabs(
+            startLine, rootAst.getColumnNumber() - 1, getTabWidth());
+    }
+
+    @Override
+    public void visitJavadocToken(DetailNode ast) {
+        // this method checks the alignment of leading asterisks.
+        if (ast.getLineNumber() != commentBeginLineNumber) {
+            final int asteriskColNumber = getAsteriskColumnNumberWithTabsExpanded(ast.getText());
+
+            checkAsteriskAlignment(
+                    ast.getLineNumber(),
+                    commentBeginColNumber,
+                    asteriskColNumber);
+        }
+    }
+
+    @Override
+    public void finishJavadocTree(DetailNode rootAst) {
+        // this method checks the alignment of closing javadoc tag.
+        final DetailAST javadocEndToken = getBlockCommentAst().getLastChild();
+
+        if (commentBeginLineNumber != javadocEndToken.getLineNo()) {
+            final String lastLine = getLastLine(javadocEndToken);
+            final int asteriskColNumber = getAsteriskColumnNumberWithTabsExpanded(lastLine);
+
+            checkAsteriskAlignment(
+                    javadocEndToken.getLineNo(),
+                    commentBeginColNumber,
+                    asteriskColNumber);
+        }
+    }
+
+    /**
+     * Processes and sets the last line of javadoc which includes comment's ending token.
+     *
+     * @param endToken javadoc ending token
+     * @return last line of javadoc comment including ending token
+     */
+    private String getLastLine(DetailAST endToken) {
+        final String lastLine;
+
+        if (endToken.getColumnNo() == 0) {
```

Comment:
Changed this, now we're getting the last line directly from using the `getLines()` method. Here is the new changes:

https://github.com/checkstyle/checkstyle/blob/dd39b32f36c619dc2ef46ca34ec21b327fc4684f/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocLeadingAsteriskAlignCheck.java#L104

https://github.com/checkstyle/checkstyle/blob/dd39b32f36c619dc2ef46ca34ec21b327fc4684f/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocLeadingAsteriskAlignCheck.java#L119-L122

https://github.com/checkstyle/checkstyle/blob/dd39b32f36c619dc2ef46ca34ec21b327fc4684f/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocLeadingAsteriskAlignCheck.java#L142-L145

We're storing the getLines() result in `fileLines` which is a global variable then we're using it to get the last line. Made the `fileLines` a global variable because it is being used in 2 places, one in beginJavadocTree() and second is finishJavadocTree()